[2] Bug 1225243 - Support triggering tasks in response to pulse messages

ydidwania opened the pull request:
I have completed the loader and client setup. 
Individual handlers for create, update and delete are still to be done

djmitche commented:
IIRC you were still working on this and not ready for review -- ping me when it is ready!

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+        await triggerHookCommon.call(this, {res, hook, payload, firedBy: 'pulseHook'});
+      });
+      for (let {exchange, routingKeyPattern} of oldBindings) {
+        await this.client.withChannel.unbindQueue(queueName, exchange, routingKeyPattern);
```
I think this (and `bindQueue`) might fail in a case where we didn't update the Queues table correctly, so we should figure out how to handle that case correctly (probably with a try/catch, and verified with a test)

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+
+    let continuationToken = undefined;
+    
+    do {
```
I think you could avoid having to deal with continuationToken here by putting the function you're passing to `forEach` into a `handler` parameter to `Hooks.scan`.

ydidwania replied:
Yes, I did that in the last couple commits. Having the `while` loop also didn't guarantee completion of the loop before the next segment (where unused queues are deleted)  is executed.

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+        if (queue) {
+          const index = this.listeners.findIndex(({_queueName}) => listener._queueName === queue.queueName);
+          if (index == -1) {
+            this.createListener(hook, queue.bindings);
```
this is async and needs to be awaited (same below)

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+    
+    // Delete the queues now left in the Queues list.
+    Queues.forEach(async (queue) => {     
+      if (this.client.withChannel.checkQueue(queue.queueName)) {
```
withChannel is a function that takes a callable, and calls it when the channel is available.

```js
await client.withChannel(async channel => { .. });
```

djmitche commented on src/main.js:
```diff
@@ -103,6 +105,53 @@ const load = loader({
+          rootUrl: cfg.taskcluster.rootUrl,
+          credentials: cfg.taskcluster.credentials,
+        }),
+        //cryptoKey: cfg.azure.cryptoKey,
```
@Biboswan found a similar thing -- this table doesn't need any encrypted columns, so no need for encryption.

ydidwania replied:
Yes, that's why I commented it :)
Will remove it in the next commit

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+    Queues.forEach(async (queue) => {     
+      if (this.client.withChannel.checkQueue(queue.queueName)) {
+        this.client.withChannel.deleteQueue(queue.queueName);
+      }
```
I think you'll need to remove the listener from `this.listeners` here, too.

djmitche replied:
This still stands.  And I didn't quite say that right -- you'll need to iterate over the listeners in `this.listeners` and call `listener.stop()` for each one.

ydidwania replied:
`this.listeners` also contains the new `listeners` created during reconciling the `Hooks` Azure table with the `Queues` Azure table. So instead of stopping all the listeners in `this.listeners`, we only have to stop the `listeners`( aka `consumers`) of the `queues`( `queueName`=`$(hookGroupId)/$(hookId)` ) which are no longer in the `Hooks` Azure table.

A queue is pulled from the `Queues` list when its corresponding `hook` is found in the `Hooks` table.  So at the end of the `do..while` loop the `Queues` list has all the queues which are no longer of use to us, and must be deleted. and the corresponding listener stopped

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+        } else {
+          this.createListener(hook, {});
+        }
+      });
```
It doesn't look like anything is ever added to the Queues table?

ydidwania commented on src/listeners.js:
```diff
@@ -0,0 +1,145 @@
+      }],
+      queueName: 'hookChanged',
+      maxLength : 50,
+      //expires : 1800000,
```
If I am deleting queues using `withChannel`, I shouldn't need to expire queues, right?

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+    // Will cancel all consumers
+    if (!this.client.isFakeClient) {
+      this.client.stop(); 
+    }
```
The PulseListener class was given a running client, so it's a little awkward for that instance to then close the client -- what if something else was also using that client?  I admit, nothing else *is* using it right now, but that might not be the case forever.

So I think that you can omit calling `this.client.stop()`.

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+    const queueName = `${hook.hookGroupId}/${hook.hookId}`; // serves as unique id for every listener
+    const listener = await pulse.consume({
+      client,      
+      queueName: queueName, 
```
you can omit the second `queueName` here

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+      const hook = this.hook;
+      // Pass empty functions in res which are executed by `triggerHookCommon`
+      // This is more of a hack so that if the payload doesnt validate against
+      // `triggerSchema`,git  we can simply ignore the message.
```
I think this comment no longer applies?

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+      }
+      for (let {exchange, routingKeyPattern} of hook.bindings) {
+        await this.client.withChannel(async channel => {channel.bindQueue(queueName, exchange, routingKeyPattern);});
+      }
```
For the case where the user adds a binding, this will in fact unbind all bindings, then re-bind them (along with the new one).  Some clever use of the `Set` class could probably avoid this, so that the only change to the queue in such a case is to call `bindQueue` for the new binding.

ydidwania replied:
I found  `_.intersectionWith(oldBindings, newBindings, _.isEqual)` and used its diff with old and new bindings to unbind and bind respectively.

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+              hookId,
+              queueName: `${hookGroupId}/${hookId}`,
+              bindings: hook.bindings,
+            });
```
This seems like something that `createListener` could handle, since it knows the queueName?

ydidwania replied:
Yes, `createListener` can do this, but we are not adding an entry to the `Queues` table everytime we are creating a new listeners. For example, when reconciling after a restart, we are only reclaiming the queues present in the `Queues` table by creating the listeners

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+            const index = this.listeners.findIndex(({_queueName}) => listener._queueName === queue.queueName);
+            if (index == -1) {
+              debug('Existing queue..creating listener');
+              await this.createListener({hook, oldBindings: queue.bindings});
```
This call will call `pulse.consume()` again, for the same queueName, meaning that there are now two listeners defined for the same queue.  Perhaps it makes sense to split up creating and deleting listeners into two functions, and separately define a function to update bindings?  Then this case can just call the update-bindings function.

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+    } while (continuationToken);
+    
+    // Delete the queues now left in the Queues list.
+    Queues.forEach(async (queue) => {     
```
For parity with `createListener`, it probably makes sense to put this in a function named `removeQueue`.

djmitche commented on src/listeners.js:
```diff
@@ -0,0 +1,180 @@
+        listener.stop();
+        this.listeners.splice(removeIndex, 1);
+      }
+      queue.remove();
```
This is an async operation and needs to be awaited!

ydidwania commented:
Yes, I forgot to include that. I think reconciling the `Hooks` table with the `Queues` and creating listeners during startup would be sufficient.

ydidwania commented:
@djmitche some missing scopes (maybe for the new Queues Azure table) are causing tests to fail. Also I had missed updating the bindings in the `Queues` Azure table when binding and unbinding exchanges from the queue

djmitche commented:
Scopes are fixed.  I'll close and re-open to test that out.  It looks like this needs to get rebased over @Biboswan's changes, too.  Otherwise, is this ready for me to have another look?

djmitche commented:
Yes, go ahead.  I'll look now (I said I'd close/re-open and meant to check back, but I forgot to do both of those things!)

djmitche commented:
32617a09bebef8c1607742f4919ac45d3275fab0 fixes the tests :)

ydidwania commented:
Haha, I thought of setting `hook.bindings` as the default, but I hadn't saved the file :)

ydidwania commented:
The other test passes locally for me.
Also `'eslint-config-taskcluster'` doesn't allow me to use decorators

djmitche commented:
Regarding the decorator, the `@` syntax is really just a fancy way of writing a function call.  So
```js
@foo
function somefunc() {
}
```
is the same as
```js
somefunc = foo(function() {
});
```

So that might be a workaround for now, if it's not a simple matter of removing a lint check.  What's the lint error?

djmitche commented:
OK, what I wrote above might be incorrect :)

https://medium.com/google-developers/exploring-es7-decorators-76ecb65fb841

ydidwania commented:
I was trying to rebase, run into a bunch of errors.
The lint error was something like
```
Unexpected Token '@'
```

djmitche commented:
OK, that sounds like a more complex fix than just adjusting the .eslintrc.  So let's use the workaround I mentioned for now, and file a bug to update eslint.

ydidwania commented:
The workaround doesn't seem to work. When I write
```
this.reconcileConsumers = serialize(async () => {
 ... 
});
```
in the `setup()` method, all  calls to `reconcileConsumers` are  skipped and tests fail because of that.

djmitche commented:
I think I got the function signature for `serialize` wrong.  I think you'll need to call it like
```js
this.reconcileConsumers = serialize(this, 'reconcileConsumers', async () => {
});
```
does that work better?

I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1514216 to try to allow actual decorators..

djmitche commented:
Hm, frustrating.  There's another approach to the same problem at
https://github.com/taskcluster/taskcluster-auth/blob/5077077f2e4094d22eef527dc9883cb8dffa387e/src/scoperesolver.js#L163-L169
which might be applicable here.  @imbstack is in the process of implementing a monorepo, after which time we can probably quickly get everything up to the same ES7 version and switch both hooks and auth to use `@serialize`.

ydidwania commented:
That worked  :)
Thats nice, some standardization will  surely help.
The only tests thats failing right now, seems to pass for me locally, is that also the case for you?

djmitche commented:
I'll check it out.

ydidwania commented:
Yes, those were extra changes that no longer had to be carried. I have removed all of those now. I also found a few trailing white spaces and removed them( there might be more).  

I factored out remove/have listeners into their own function in the latest commit which makes things a bit more readable, but you are right there is more that can be done. It is absolutely fine with me if you do those. Go ahead :)

ydidwania commented:
@djmitche I rebased all the commits and then I had to make some changes because of the version upgrade of `eslint` in [eslint-config-taskcluster](https://github.com/taskcluster/eslint-config-taskcluster). 
Also I was running into an error because of caching of injected components, so I changed the `withPulse` in `test/helper.js` to use the already injected `Taskcreator` component instead of loading it again.

djmitche commented:
Sounds great -- thanks for the notes on the rebase.
